pkg/util/util.go

```diff
@@ -466,3 +463,54 @@ func GetTopologySpreadConstraintsFromNodes(
 	}
 	return constraints, nil
 }
+
+// UpdateStorageClusterCondition update condition based on source and type
+func UpdateStorageClusterCondition(
+	cluster *corev1.StorageCluster,
+	toUpdate *corev1.ClusterCondition,
+) {
+	if toUpdate == nil || toUpdate.Source == "" || toUpdate.Type == "" {
+		logrus.Warn("empty or invalid storage cluster condition provided")
+		return
+	}
+
+	// Search for condition by source and type
+	foundIndex := -1
+	for i, condition := range cluster.Status.Conditions {
+		if toUpdate.Source == condition.Source && toUpdate.Type == condition.Type {
+			foundIndex = i
+			// No update needed but populate the last transition time
+			if toUpdate.Status == condition.Status &&
+				toUpdate.Message == condition.Message {
+				toUpdate.LastTransitionTime = condition.LastTransitionTime
+				return
+			}
+			break
+		}
+	}
+
+	// Create a new condition or overwrite existing one
+	if toUpdate.LastTransitionTime.Equal(&metav1.Time{}) {
+		// Set timestamp if not provided, truncate to second
+		toUpdate.LastTransitionTime = metav1.NewTime(time.Now().Truncate(time.Second))
+	}
+	if foundIndex == -1 {
+		cluster.Status.Conditions = append(cluster.Status.Conditions, *toUpdate)
+	} else {
+		cluster.Status.Conditions[foundIndex] = *toUpdate.DeepCopy()
```

Comment:
this should be moved to the first element, otherwise we won't know what's the latest condition when we do kubectl describe.